### ISSUE TITLE
Log and indexing do not commute correctly with respect to gradient

### ISSUE THREAD

**Issue by ae-foster** [pytorch/pytorch#12986]
##  Bug
I encountered an unexpected `nan` gradient that was rectified by swapping the order of `torch.log` and an indexing operation `[a > 0.]`. These ought to commute right

## To Reproduce

```
>>> import torch
>>>
>>> a = torch.tensor([0., 1., 2.], requires_grad=True)
>>> l = torch.log(a)[a > 0.].sum()
>>> l.backward()
>>> print(a.grad)
tensor([   nan, 1.0000, 0.5000])
>>> a.grad.data.zero_()
tensor([0., 0., 0.])
>>> l = torch.log(a[a > 0.]).sum()
>>> l.backward()
>>> print(a.grad)
tensor([0.0000, 1.0000, 0.5000])
```

## Expected behavior

The second answer `tensor([0.0000, 1.0000, 0.5000])` is the correct gradient, since `a[0]` is not used in calculating `l`.

## Environment

PyTorch version: 0.4.0 (also tested on 0.4.1/CPU)
Is debug build: No
CUDA used to build PyTorch: 9.1.85

OS: Fedora release 28 (Twenty Eight)
GCC version: (GCC) 8.2.1 20181011 (Red Hat 8.2.1-4)
CMake version: version 3.11.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy (1.14.5)
[pip3] torch (0.4.0)
[pip3] torchfile (0.1.0)
[pip3] torchvision (0.2.1)
[conda] Could not collect

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

**Comment by fmassa** [comment]
@weiyangfb a possible fix would be to return sparse gradients from indexing operators. But we should give a careful thought on this to see if it would be more trouble than what it is worth it.

**Comment by weiyangfb** [comment]
@fmassa sounds reasonable to me as long as `to_dense()` is relatively efficient. Another thing is, a sparse gradient might be slower than a dense one unless the gradient is large and sparse enough.

**Comment by apaszke** [comment]
@fmassa returning sparse gradients will likely blow up all other autograd functions, which are written with the assumption that their inputs are dense.

**Comment by albanD** [comment]
> These ought to commute right

Since you're at a point where log is not continuously differentiable, you cannot expect them to commute I'm afraid.

The behavior you're seeing is expected and as mentionned above, the only way to fix that would be to be able to see the difference between 0 grad and not defined grads which we cannot do with current Tensors.
